### ISSUE TITLE
Gradient accumulation error with optax (jax.lax.cond)

### ISSUE THREAD

**Issue by zw123han** [alpa-projects/alpa#839]
Hello. It seems Alpa errors out with jax.lax.cond, which is required for gradient accumulation using Optax.MultiSteps https://optax.readthedocs.io/en/latest/api.html#optax.MultiSteps

optax.MultiSteps internally uses jax.lax.cond which caused the error. https://github.com/deepmind/optax/blob/master/optax/_src/wrappers.py#L271#L427
![image](https://user-images.githubusercontent.com/67124639/210034934-9ec6aede-652c-4384-979d-d8ec6961dc9b.png)

Is there any other method to perform gradient accumulation or any ideas on the source of this issue?

### Reproduce

Replace optimizer with the wrapped optimizer for gradient accumulation.

`optimizer = optax.MultiSteps(optax.adamw( `
                `learning_rate=linear_decay_lr_schedule_fn,`
                `b1=training_args.adam_beta1,`
                `b2=training_args.adam_beta2,`
                `eps=training_args.adam_epsilon,`
                `weight_decay=training_args.weight_decay,`
               ` mask=decay_mask_fn), every_k_schedule=10) ` in https://github.com/alpa-projects/alpa/blob/main/examples/opt_finetune/run_clm_flax.py

### Stack Trace
source_file="/h/hanziwen/.local/lib/python3.9/site-packages/optax/_src/wrappers.py" source_line=395}
*** SIGABRT received at time=1672374756 on cpu 31 ***
PC: @     0x7f4f5f4cee87  (unknown)  raise
    @     0x7f4f5f893980  (unknown)  (unknown)
    @     0x7f4e948e8270       2320  xla::spmd::BuildInstructionBatchDimMap()
    @     0x7f4e95ff9c6b       2352  xla::spmd::AutoSharding::Run()
    @     0x7f4e9888d1e1        816  xla::HloPassPipeline::RunPassesInternal<>()
    @     0x7f4e9888de35        448  xla::HloPassPipeline::Run()
    @     0x7f4e94e5ebd4         80  xla::HloPassInterface::Run()
    @     0x7f4e94e68041       4128  xla::spmd::RunAutoShardingPass()
    @     0x7f4e94e53e1a        160  pybind11::cpp_function::initialize<>()::{lambda()#3}::_FUN()
    @     0x7f4e94cc42c0        576  pybind11::cpp_function::dispatcher()
    @     0x56097204de14  (unknown)  cfunction_call
    @ ... and at least 8 more frames
[2022-12-29 23:32:36,416 E 23762 23762] logging.cc:361: *** SIGABRT received at time=1672374756 on cpu 31 ***
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361: PC: @     0x7f4f5f4cee87  (unknown)  raise
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4f5f893980  (unknown)  (unknown)
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4e948e8270       2320  xla::spmd::BuildInstructionBatchDimMap()
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4e95ff9c6b       2352  xla::spmd::AutoSharding::Run()
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4e9888d1e1        816  xla::HloPassPipeline::RunPassesInternal<>()
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4e9888de35        448  xla::HloPassPipeline::Run()
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4e94e5ebd4         80  xla::HloPassInterface::Run()
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4e94e68041       4128  xla::spmd::RunAutoShardingPass()
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4e94e53e1a        160  pybind11::cpp_function::initialize<>()::{lambda()#3}::_FUN()
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x7f4e94cc42c0        576  pybind11::cpp_function::dispatcher()
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @     0x56097204de14  (unknown)  cfunction_call
[2022-12-29 23:32:36,417 E 23762 23762] logging.cc:361:     @ ... and at least 8 more frames
Fatal Python error: Aborted

Stack (most recent call first):
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/shard_parallel/auto_sharding.py", line 344 in run_auto_sharding_pass
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/shard_parallel/compile_executable.py", line 169 in shard_parallel_internal_gradient_accumulation
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/shard_parallel/compile_executable.py", line 82 in compile_shard_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/parallel_method.py", line 105 in compile_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/api.py", line 218 in _compile_parallel_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/jax/linear_util.py", line 309 in memoized_fun
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/api.py", line 191 in _decode_args_and_get_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/api.py", line 127 in get_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/create_state_parallel.py", line 88 in compile_create_state_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/parallel_method.py", line 356 in compile_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/api.py", line 218 in _compile_parallel_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/jax/linear_util.py", line 309 in memoized_fun
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/api.py", line 191 in _decode_args_and_get_executable
  File "/h/hanziwen/.local/lib/python3.9/site-packages/alpa/api.py", line 121 in __call__
  File "/h/hanziwen/.local/lib/python3.9/site-packages/jax/_src/traceback_util.py", line 162 in reraise_with_filtered_traceback
  File "/h/hanziwen/opt-experiments/examples/opt_finetune/run_clm_flax_grad_acc.py", line 914 in main
  File "/h/hanziwen/opt-experiments/examples/opt_finetune/run_clm_flax_grad_acc.py", line 1093 in <module>

**Comment by zw123han** [comment]
Update: I implemented my own gradient accumulation optimizer without the need to use jax.lax.cond based on the implementation of optax.apply_every (https://github.com/deepmind/optax/blob/master/optax/_src/transform.py#L876#L909), but ensuring equivalence to optax.MultiSteps.  
However, this issue remains for those who want want to use the default implementation. Feel free to close.
